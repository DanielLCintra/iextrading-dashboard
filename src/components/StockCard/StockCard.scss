.StockCard {
  margin-bottom: 15px;
}
