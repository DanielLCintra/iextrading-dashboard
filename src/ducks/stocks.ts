import { AnyAction, Dispatch } from 'redux';
import Api from '../utils/Api';

export const DUCK_NAME = 'stocks';

// Interfaces

export interface IStock {
  symbol: string,
  companyName: string,
  sector: string,
  latestPrice: number,
  ytdChange: number,
}

export interface IStocksState {
  data: IStock[] | null,
  error: Error | null,
  loading: boolean,
}

// Action types

export const LOAD_STOCKS_STARTED = `${DUCK_NAME}/LOAD_STOCKS_STARTED`;
interface ILoadStocksStarted {
  type: typeof LOAD_STOCKS_STARTED,
}

export const LOAD_STOCKS_SUCCEED = `${DUCK_NAME}/LOAD_STOCKS_SUCCEED`;
interface ILoadStocksSucceed {
  type: typeof LOAD_STOCKS_SUCCEED,
  data: IStock[],
}

export const LOAD_STOCKS_FAILED = `${DUCK_NAME}/LOAD_STOCKS_FAILED`;
interface ILoadStocksFailed {
  type: typeof LOAD_STOCKS_FAILED,
  error: Error,
}

// Action creators

export const loadStocksStarted = ():ILoadStocksStarted =>
  ({ type: LOAD_STOCKS_STARTED });

export const loadStocksSucceed = (data:IStock[]):ILoadStocksSucceed =>
  ({ type: LOAD_STOCKS_SUCCEED, data });

export const loadStocksFailed = (error:Error):ILoadStocksFailed =>
  ({ type: LOAD_STOCKS_FAILED, error });

// Thunks

export const loadStocks = () =>
  async (dispatch:Dispatch) => {
    dispatch(loadStocksStarted());
    try {
      const results = await Api.instance.getStocks();
      dispatch(loadStocksSucceed(results.data));
      return Promise.resolve(results);
    } catch (error) {
      dispatch(loadStocksFailed(error));
      return Promise.reject(error);
    }
  };

// Reducer

export const INITIAL_STATE:IStocksState = {
  data: null,
  error: null,
  loading: false,
};

const reducer = (state = INITIAL_STATE, action:AnyAction) => {
  switch (action.type) {
  case LOAD_STOCKS_STARTED:
    return {
      ...state,
      loading: true,
    };
  case LOAD_STOCKS_SUCCEED:
    return {
      ...state,
      data: [
        ...(state.data || []),
        ...action.data,
      ],
      loading: false,
    };
  case LOAD_STOCKS_FAILED:
    return {
      ...state,
      error: action.error,
      loading: false,
    };
  default: return state;
  }
};

export default reducer;
