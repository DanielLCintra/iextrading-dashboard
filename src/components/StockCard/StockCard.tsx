import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import * as React from 'react';
import { Link } from 'react-router-dom'
import { IStock } from '../../ducks/stocks';
import classes from './StockCard.scss';

export interface IStockCardProps {
  stock: IStock,
};

class StockCard extends React.Component<any, IStockCardProps> {
  public render() {
    const stock = this.props.stock;
    return (
      <Link
        to={`/stock?symbol=${stock.symbol}`}
        className={classes.StockCardLink}
      >
        <Card className={classes.StockCard}>
          <CardContent>
            <Typography
              className={classes.StockCardCompanyName}
              color="textSecondary"
            >
              {stock.companyName}
            </Typography>
            <Typography
              className={classes.StockCardCode}
              variant="headline"
              component="h2"
            >
              {stock.symbol}
            </Typography>
            <Typography
              className={classes.StockCardSector}
              color="textSecondary"
            >
              {stock.sector}
            </Typography>
          </CardContent>
        </Card>
      </Link>
    );
  }
}

export default StockCard;
