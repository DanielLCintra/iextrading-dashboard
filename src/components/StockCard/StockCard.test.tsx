import { mount } from 'enzyme';
import * as React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Link } from 'react-router-dom';
import { createFakeStock } from '../../ducks/stocks.test';
import StockCard, { IStockCardProps } from './StockCard';

describe('StockCard', () => {
  const props:IStockCardProps = {
    stock: createFakeStock(),
  };

  const wrapper = mount(
    <MemoryRouter>
      <StockCard {...props} />
    </MemoryRouter>
  );

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Link for the detail', () => {
    expect(wrapper.find(Link).length).toBeGreaterThan(0);
  });
});
