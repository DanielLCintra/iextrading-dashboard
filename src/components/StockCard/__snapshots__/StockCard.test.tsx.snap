// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockCard renders without crashing 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "4gttwb",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "4gttwb",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <StockCard
      stock={
        Object {
          "companyName": "Testing Co",
          "latestPrice": 0,
          "sector": "Tests",
          "symbol": "TST",
          "ytdChange": 0,
        }
      }
    >
      <Link
        replace={false}
        to="/stock/TST"
      >
        <a
          href="/stock/TST"
          onClick={[Function]}
        >
          <WithStyles(Card)>
            <Card
              classes={
                Object {
                  "root": "MuiCard-root-1",
                }
              }
              raised={false}
            >
              <WithStyles(Paper)
                className="MuiCard-root-1"
                elevation={1}
              >
                <Paper
                  className="MuiCard-root-1"
                  classes={
                    Object {
                      "elevation0": "MuiPaper-elevation0-4",
                      "elevation1": "MuiPaper-elevation1-5",
                      "elevation10": "MuiPaper-elevation10-14",
                      "elevation11": "MuiPaper-elevation11-15",
                      "elevation12": "MuiPaper-elevation12-16",
                      "elevation13": "MuiPaper-elevation13-17",
                      "elevation14": "MuiPaper-elevation14-18",
                      "elevation15": "MuiPaper-elevation15-19",
                      "elevation16": "MuiPaper-elevation16-20",
                      "elevation17": "MuiPaper-elevation17-21",
                      "elevation18": "MuiPaper-elevation18-22",
                      "elevation19": "MuiPaper-elevation19-23",
                      "elevation2": "MuiPaper-elevation2-6",
                      "elevation20": "MuiPaper-elevation20-24",
                      "elevation21": "MuiPaper-elevation21-25",
                      "elevation22": "MuiPaper-elevation22-26",
                      "elevation23": "MuiPaper-elevation23-27",
                      "elevation24": "MuiPaper-elevation24-28",
                      "elevation3": "MuiPaper-elevation3-7",
                      "elevation4": "MuiPaper-elevation4-8",
                      "elevation5": "MuiPaper-elevation5-9",
                      "elevation6": "MuiPaper-elevation6-10",
                      "elevation7": "MuiPaper-elevation7-11",
                      "elevation8": "MuiPaper-elevation8-12",
                      "elevation9": "MuiPaper-elevation9-13",
                      "root": "MuiPaper-root-2",
                      "rounded": "MuiPaper-rounded-3",
                    }
                  }
                  component="div"
                  elevation={1}
                  square={false}
                >
                  <div
                    className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
                  >
                    <WithStyles(CardContent)>
                      <CardContent
                        classes={
                          Object {
                            "root": "MuiCardContent-root-29",
                          }
                        }
                        component="div"
                      >
                        <div
                          className="MuiCardContent-root-29"
                        >
                          <WithStyles(Typography)
                            color="textSecondary"
                          >
                            <Typography
                              align="inherit"
                              classes={
                                Object {
                                  "alignCenter": "MuiTypography-alignCenter-43",
                                  "alignJustify": "MuiTypography-alignJustify-45",
                                  "alignLeft": "MuiTypography-alignLeft-42",
                                  "alignRight": "MuiTypography-alignRight-44",
                                  "body1": "MuiTypography-body1-39",
                                  "body2": "MuiTypography-body2-38",
                                  "button": "MuiTypography-button-41",
                                  "caption": "MuiTypography-caption-40",
                                  "colorError": "MuiTypography-colorError-54",
                                  "colorInherit": "MuiTypography-colorInherit-49",
                                  "colorPrimary": "MuiTypography-colorPrimary-50",
                                  "colorSecondary": "MuiTypography-colorSecondary-51",
                                  "colorTextPrimary": "MuiTypography-colorTextPrimary-52",
                                  "colorTextSecondary": "MuiTypography-colorTextSecondary-53",
                                  "display1": "MuiTypography-display1-34",
                                  "display2": "MuiTypography-display2-33",
                                  "display3": "MuiTypography-display3-32",
                                  "display4": "MuiTypography-display4-31",
                                  "gutterBottom": "MuiTypography-gutterBottom-47",
                                  "headline": "MuiTypography-headline-35",
                                  "noWrap": "MuiTypography-noWrap-46",
                                  "paragraph": "MuiTypography-paragraph-48",
                                  "root": "MuiTypography-root-30",
                                  "subheading": "MuiTypography-subheading-37",
                                  "title": "MuiTypography-title-36",
                                }
                              }
                              color="textSecondary"
                              gutterBottom={false}
                              headlineMapping={
                                Object {
                                  "body1": "p",
                                  "body2": "aside",
                                  "display1": "h1",
                                  "display2": "h1",
                                  "display3": "h1",
                                  "display4": "h1",
                                  "headline": "h1",
                                  "subheading": "h3",
                                  "title": "h2",
                                }
                              }
                              noWrap={false}
                              paragraph={false}
                              variant="body1"
                            >
                              <p
                                className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-53"
                              >
                                Testing Co
                              </p>
                            </Typography>
                          </WithStyles(Typography)>
                          <WithStyles(Typography)
                            component="h2"
                            variant="headline"
                          >
                            <Typography
                              align="inherit"
                              classes={
                                Object {
                                  "alignCenter": "MuiTypography-alignCenter-43",
                                  "alignJustify": "MuiTypography-alignJustify-45",
                                  "alignLeft": "MuiTypography-alignLeft-42",
                                  "alignRight": "MuiTypography-alignRight-44",
                                  "body1": "MuiTypography-body1-39",
                                  "body2": "MuiTypography-body2-38",
                                  "button": "MuiTypography-button-41",
                                  "caption": "MuiTypography-caption-40",
                                  "colorError": "MuiTypography-colorError-54",
                                  "colorInherit": "MuiTypography-colorInherit-49",
                                  "colorPrimary": "MuiTypography-colorPrimary-50",
                                  "colorSecondary": "MuiTypography-colorSecondary-51",
                                  "colorTextPrimary": "MuiTypography-colorTextPrimary-52",
                                  "colorTextSecondary": "MuiTypography-colorTextSecondary-53",
                                  "display1": "MuiTypography-display1-34",
                                  "display2": "MuiTypography-display2-33",
                                  "display3": "MuiTypography-display3-32",
                                  "display4": "MuiTypography-display4-31",
                                  "gutterBottom": "MuiTypography-gutterBottom-47",
                                  "headline": "MuiTypography-headline-35",
                                  "noWrap": "MuiTypography-noWrap-46",
                                  "paragraph": "MuiTypography-paragraph-48",
                                  "root": "MuiTypography-root-30",
                                  "subheading": "MuiTypography-subheading-37",
                                  "title": "MuiTypography-title-36",
                                }
                              }
                              color="default"
                              component="h2"
                              gutterBottom={false}
                              headlineMapping={
                                Object {
                                  "body1": "p",
                                  "body2": "aside",
                                  "display1": "h1",
                                  "display2": "h1",
                                  "display3": "h1",
                                  "display4": "h1",
                                  "headline": "h1",
                                  "subheading": "h3",
                                  "title": "h2",
                                }
                              }
                              noWrap={false}
                              paragraph={false}
                              variant="headline"
                            >
                              <h2
                                className="MuiTypography-root-30 MuiTypography-headline-35"
                              >
                                TST
                              </h2>
                            </Typography>
                          </WithStyles(Typography)>
                          <WithStyles(Typography)
                            color="textSecondary"
                          >
                            <Typography
                              align="inherit"
                              classes={
                                Object {
                                  "alignCenter": "MuiTypography-alignCenter-43",
                                  "alignJustify": "MuiTypography-alignJustify-45",
                                  "alignLeft": "MuiTypography-alignLeft-42",
                                  "alignRight": "MuiTypography-alignRight-44",
                                  "body1": "MuiTypography-body1-39",
                                  "body2": "MuiTypography-body2-38",
                                  "button": "MuiTypography-button-41",
                                  "caption": "MuiTypography-caption-40",
                                  "colorError": "MuiTypography-colorError-54",
                                  "colorInherit": "MuiTypography-colorInherit-49",
                                  "colorPrimary": "MuiTypography-colorPrimary-50",
                                  "colorSecondary": "MuiTypography-colorSecondary-51",
                                  "colorTextPrimary": "MuiTypography-colorTextPrimary-52",
                                  "colorTextSecondary": "MuiTypography-colorTextSecondary-53",
                                  "display1": "MuiTypography-display1-34",
                                  "display2": "MuiTypography-display2-33",
                                  "display3": "MuiTypography-display3-32",
                                  "display4": "MuiTypography-display4-31",
                                  "gutterBottom": "MuiTypography-gutterBottom-47",
                                  "headline": "MuiTypography-headline-35",
                                  "noWrap": "MuiTypography-noWrap-46",
                                  "paragraph": "MuiTypography-paragraph-48",
                                  "root": "MuiTypography-root-30",
                                  "subheading": "MuiTypography-subheading-37",
                                  "title": "MuiTypography-title-36",
                                }
                              }
                              color="textSecondary"
                              gutterBottom={false}
                              headlineMapping={
                                Object {
                                  "body1": "p",
                                  "body2": "aside",
                                  "display1": "h1",
                                  "display2": "h1",
                                  "display3": "h1",
                                  "display4": "h1",
                                  "headline": "h1",
                                  "subheading": "h3",
                                  "title": "h2",
                                }
                              }
                              noWrap={false}
                              paragraph={false}
                              variant="body1"
                            >
                              <p
                                className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-53"
                              >
                                Tests
                              </p>
                            </Typography>
                          </WithStyles(Typography)>
                        </div>
                      </CardContent>
                    </WithStyles(CardContent)>
                  </div>
                </Paper>
              </WithStyles(Paper)>
            </Card>
          </WithStyles(Card)>
        </a>
      </Link>
    </StockCard>
  </Router>
</MemoryRouter>
`;
